### PR TITLE
initial commit for newtabprefprovider

### PR DESCRIPTION
**Pull request by sarracini**

### CODE REVIEW

**Comment by sarracini** [comment]
@oyiptong r?

File: browser/components/newtab/NewTabPrefsProvider.jsm [hunk 1]
```diff
@@ -31,7 +31,24 @@ PrefsProvider.prototype = {
   observe(subject, topic, data) { // jshint ignore:line
     if (topic === "nsPref:changed") {
       if (prefsSet.has(data)) {
-        this.emit(data);
+        switch (data) {
+          case "browser.newtabpage.enabled":
+            this.emit("setEnabled", {
```
Comment (oyiptong) [inline browser/components/newtab/NewTabPrefsProvider.jsm]: Can you please use the name of the pref as the message it emits?

It makes for good semantics: you listen "on" a pref

i.e. you'd do NewTabPrefsProvider.on("browser.newtabpage.enabled") as opposed to NewTabPrefsProvider.on("setEnabled")

Comment (oyiptong) [inline browser/components/newtab/NewTabPrefsProvider.jsm]: you also don't need to return an object. You could just return the value.

File: browser/components/newtab/NewTabPrefsProvider.jsm [hunk 2]
```diff
@@ -31,7 +31,24 @@ PrefsProvider.prototype = {
...
+              enabled: Services.prefs.getBoolPref(data),
+            });
+            break;
+          case "browser.newtabpage.enhanced":
```
Comment (oyiptong) [inline browser/components/newtab/NewTabPrefsProvider.jsm:38]: ditto here.

You could replace the switch/case statement by an if clause.

``` javascript
if (data === "browser.newtabpage.enabled" || data ===" browser.newtabpage.enhanced") {
  this.emit(data, Services.prefs.getBoolPref(data));
} else {
  this.emit(data)
}
```

File: browser/components/newtab/RemoteAboutNewTab.jsm [hunk 3]
```diff
@@ -331,6 +343,14 @@ let RemoteAboutNewTab = {
     }
   },
 
+  setEnabled: function(name, data) {
+    this.pageListener.sendAsyncMessage("NewTab:setEnabled", data.enabled);
```
Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm]: following the suggested changes above, in NewTabPrefsProvider.jsm, you wouldn't need to unpack any object

File: browser/components/newtab/RemoteAboutNewTab.jsm [hunk 4]
```diff
@@ -331,6 +343,14 @@ let RemoteAboutNewTab = {
...
+  },
+
+  setEnhanced: function(name, data) {
```
Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm:359]: ditto here

File: browser/components/nsBrowserGlue.js [hunk 5]
```diff
@@ -1173,6 +1173,7 @@ BrowserGlue.prototype = {
     WebappManager.uninit();
 
     RemoteAboutNewTab.uninit();
+    NewTabPrefsProvider.prefs.stopTracking();
```
Comment (oyiptong) [inline browser/components/nsBrowserGlue.js:1176]: :+1:

**Comment by oyiptong** [comment]
note that the changes you made in NewTabPrefsProvider would make the xpcshell test fail.

the test is listening `on.("browser.newtabpage.enabled", callback)`.

Don't forget to run tests ;-)

Also, please modify/add tests to verify that data is obtained in the callbacks for prefs that emit them.

File: browser/components/newtab/NewTabPrefsProvider.jsm (hunk 2 above)
Comment (sarracini) [inline browser/components/newtab/NewTabPrefsProvider.jsm:38]: i used a switch case that way if there are other prefs in the future that need to be added to newtab it's easy to add them in

Comment (oyiptong) [inline browser/components/newtab/NewTabPrefsProvider.jsm:38]: I'm fine either way. One way to see it is... one approach has less code than the other ;-)

Comment (sarracini) [inline browser/components/newtab/NewTabPrefsProvider.jsm:38]: thoughts on: 

``` javascript
 switch (data) {
      case "browser.newtabpage.enabled":
      case "browser.newtabpage.enhanced":
        this.emit(data, Services.prefs.getBoolPref(data));
        break;
```

Comment (sarracini) [inline browser/components/newtab/NewTabPrefsProvider.jsm:38]: ahhh nevermind that won't work!

File: browser/components/newtab/RemoteAboutNewTab.jsm [hunk 6]
```diff
@@ -343,6 +363,8 @@ let RemoteAboutNewTab = {
     PlacesProvider.links.on("clearHistory", this.placesClearHistory.bind(this));
     PlacesProvider.links.on("linkChanged", this.placesLinkChanged.bind(this));
     PlacesProvider.links.on("manyLinksChanged", this.placesManyLinksChanged.bind(this));
+    NewTabPrefsProvider.prefs.on("browser.newtabpage.enabled", this.setEnabled.bind(this));
+    NewTabPrefsProvider.prefs.on("browser.newtabpage.enhanced", this.setEnhanced.bind(this));
```
Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm:380]: we should also listen to the `pinned` change notification

Comment (sarracini) [inline browser/components/newtab/RemoteAboutNewTab.jsm:380]: what's the callback for pinned?

File: browser/components/newtab/NewTabPrefsProvider.jsm (hunk 2 above)
Comment (oyiptong) [inline browser/components/newtab/NewTabPrefsProvider.jsm:38]: > thoughts on:
> 
>  switch (data) {
>       case "browser.newtabpage.enabled":
>       case "browser.newtabpage.enhanced":
>         this.emit(data, Services.prefs.getBoolPref(data));
>         break;
> —
> 
> IMO: better be explicit than being clever

File: browser/components/newtab/RemoteAboutNewTab.jsm (hunk 6 above)
Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm:380]: we need to make one. We can send the message to the page, and write an issue on mozilla/remote-newtab to handle it

Comment (oyiptong) [inline browser/components/newtab/RemoteAboutNewTab.jsm:380]: here we go: https://github.com/mozilla/remote-newtab/issues/137

**Comment by oyiptong** [comment]
r+

**Pull request merged**